plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/AddFilesProcedure.java

```diff
@@ -0,0 +1,403 @@
+            }
+            storageFormat = extractHiveStorageFormat(sourceTable.getStorage().getStorageFormat());
+            sourceLocation = sourceTable.getStorage().getLocation();
+            sourceType = SourceType.TABLE;
+        }
+        else if (location != null && format != null) {
+            checkProcedureArgument(sourceSchemaName == null, "SOURCE_SCHEMA_NAME argument must be null");
+            checkProcedureArgument(sourceTableName == null, "SOURCE_TABLE_NAME argument must be null");
+            checkProcedureArgument(partitionFilter == null, "PARTITION_FILTER argument must be null");
+
+            sourceLocation = location;
+            storageFormat = HiveStorageFormat.valueOf(format.toUpperCase(ENGLISH));
+            checkProcedureArgument(storageFormat == ORC || storageFormat == PARQUET || storageFormat == AVRO, "The procedure does not support storage format: %s", storageFormat);
+            sourceType = SourceType.LOCATION;
+        }
+        else {
+            throw new TrinoException(INVALID_PROCEDURE_ARGUMENT, "Either source schema and table names, or location and format must be provided");
+        }
+
+        try {
+            TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+            ImmutableList.Builder<DataFile> dataFilesBuilder = ImmutableList.builder();
+            if (!partitionSpec.isPartitioned()) {
+                log.debug("Building data files from %s", sourceLocation);
+                dataFilesBuilder.addAll(buildDataFiles(fileSystem, sourceType, recursive, storageFormat, sourceLocation, partitionSpec, Optional.empty(), schema, requiredFields));
+            }
+            else {
+                checkProcedureArgument(location == null, "The procedure does not support partitioned tables with location argument");
+
+                Map<String, Optional<Partition>> partitions = metastore.getPartitionsByNames(sourceTable, firstNonNull(partitionFilter, ImmutableList.of()));
+                int fileCount = 1;
+                for (Map.Entry<String, Optional<Partition>> partition : partitions.entrySet()) {
+                    Storage storage = partition.getValue().orElseThrow(() -> new IllegalArgumentException("Invalid partition: " + partition.getKey())).getStorage();
+                    log.debug("Building data files from '%s' for partition %d of %d", storage.getLocation(), fileCount++, partitions.size());
+                    HiveStorageFormat partitionStorageFormat = extractHiveStorageFormat(storage.getStorageFormat());
+                    StructLike partitionData = DataFiles.data(partitionSpec, partition.getKey());
+                    dataFilesBuilder.addAll(buildDataFiles(fileSystem, sourceType, recursive, partitionStorageFormat, storage.getLocation(), partitionSpec, Optional.of(partitionData), schema, requiredFields));
+                }
+            }
+
+            log.debug("Start new transaction");
+            Transaction transaction = table.newTransaction();
```

Comment:
The equivalent iceberg library procedure also does a check for duplicate files (`check_duplicate_files`), are we sure we don't need that ?

Reply:
I hope we will handle it in a follow-up as I already left a comment in tests. Let me add additional comments here and document the limitation.